src/content/developers/tutorials/merkle-proofs-for-offline-data-integrity/index.md

```diff
@@ -0,0 +1,253 @@
+and price. However, integrity is usually assured.
+
+In this article you learn **how** to ensure data integrity without storing the data on the blockchain, using
+[Merkle proofs](https://computersciencewiki.org/index.php/Merkle_proof). 
+
+## How does it work?
+
+In theory we could just store the hash of the data on chain, and send all the data in transactions that require it. However, this is still too expensive. A byte of data to a transaction costs about 16 gas, currently about half a cent, or about $5 per kilobyte. At $5000 per megabyte, this is still too expensive for many uses, even without the added cost of hashing the data.
+
+The solution is to repeatedly hash different subsets of the data, so for the data that you don't need to send you can just send a hash. You do this using a Merkle tree, a tree data structure where each node is a hash of the nodes below it:
+
+![Merkle Tree](tree.png)
+
+The root hash is the only part that needs to be stored on chain. To prove a certain value, you provide all the hashes that need to be combined with it to obtain the root. For example, to prove `C` you provide `D`, `H(A-B)`, and `H(E-H)`.
+
+![Proof of the value of C](proof-c.png)
+
+
+## Implementation
+
+[The sample code is provided here](https://github.com/qbzzt/merkle-proofs-for-offline-data-integrity).
+
+### Off-chain code
+
+In this article we use JavaScript for the off-chain computations. Most decentralized applications have their off-chain component in JavaScript.
+
+#### Creating the Merkle root
+
+First we need to provide the Merkle root to the chain.
+
+```javascript
+const ethers = require('ethers')
+```
+
+[We use the hash function from the ethers package](https://docs.ethers.io/v5/api/utils/hashing/#utils-keccak256).
+
+```javascript
+// The raw data whose integrity we have to verify. The first two bytes a
+// are a user identifier, and the last two bytes the amount of tokens the
+// user owns at present.
+const dataArray = [
+    0x0BAD0010, 0x60A70020, 0xBEEF0030, 0xDEAD0040, 0xCA110050,
+    0x0E660060, 0xFACE0070, 0xBAD00080, 0x060D0091
+]
+```
+
+Encoding each entry into a single 256 bit integer results in less readable code than using JSON, for example. However, this means significantly less processing to retrieve the data in the contract, so much lower gas costs. [You can read JSON on chain](https://github.com/chrisdotn/jsmnSol), it's just a bad idea if avoidable.
+
+```javascript
+// The array of hash values, as BigInts 
+const hashArray = dataArray; 
+```
+
+In this case our data is 256 bit values to begin with, so no processing is needed. If we use a more complicated data structure, such as strings, we need to make sure we hash the data first to get an array of hashes. Note that this is also because we don't care if users know other users' information. Otherwise we would have had to hash so user 1 won't know the value for user 0, user 2 won't know the value for user 3, etc.
```

Comment:
```suggestion
In this case our data is 256-bit values to begin with, so no processing is needed. If we use a more complicated data structure, such as strings, we need to make sure we hash the data first to get an array of hashes. Note that this is also because we don't care if users know other users' information. Otherwise we would have had to hash so user 1 won't know the value for user 0, user 2 won't know the value for user 3, etc.
```